tensorstore/internal/http/http_request.h

```diff
@@ -88,19 +101,21 @@ class HttpRequestBuilder {
   /// response.
   HttpRequestBuilder& EnableAcceptEncoding();
 
+  /// Adds a `range` header to the http request if the byte_range
+  /// is specified.
+  HttpRequestBuilder& AddRangeHeader(OptionalByteRangeRequest byte_range, bool & result);
```

Comment:
I'd rather not have an output parameter on the builder, particularly one which is ignored. I prefer that we do the following:

1. Where the bool is intended to indicate whether it was valid, we should constrain the caller.
    If, instead, we're just going to ignore the error, rename the method to `MaybeAddRangeHeader` and similar.
2. And/or on a non-ignroable failure (probably none of these header methods?), store an internal status in the `HttpRequestBuilder` object, and change Build to return `Result<>`, propagating an error status.

Also, looking at the S3 Request builder, it seems like there are some ordering requirements on the header order. I think that the request builder could easily be changed to store a map/multimap of header -> value, much like is done in the response. Then the S3RequestBuilder can construct the digest from that without additional manipulation.

Reply:
> 1. Where the bool is intended to indicate whether it was valid, we should constrain the caller.
>    If, instead, we're just going to ignore the error, rename the method to `MaybeAddRangeHeader` and similar.

- `AddRangeHeader` -> `MaybeAddRangeHeader` 
- `AddCacheControlMaxAgeHeader` -> `MaybeAddCacheControlMaxAgeHeader`
- `AddStalenessBoundCacheControlHeader` -> `MaybeAddStalenessBoundCacheControlHeader`

> 2. And/or on a non-ignroable failure (probably none of these header methods?), store an internal status in the `HttpRequestBuilder` object, and change Build to return `Result<>`, propagating an error status.

Done.

> Also, looking at the S3 Request builder, it seems like there are some ordering requirements on the header order. I think that the request builder could easily be changed to store a map/multimap of header -> value, much like is done in the response. Then the S3RequestBuilder can construct the digest from that without additional manipulation.

Done.